torch/csrc/jit/passes/create_autodiff_subgraphs.cpp

```diff
@@ -12,6 +13,66 @@ namespace jit {
 
 namespace {
 
+std::vector<c10::optional<const Use>> gatherLastUses(
+    at::ArrayRef<Value*> values) {
+  return fmap(values, [&](Value* v) -> c10::optional<const Use> {
+    return firstOrLastUse(v, /*find_first*/ false);
+  });
+}
+
+// When merging a node into a subgraph, we wish to preserve all of the
+// aliasing properties of the node's outputs. It is difficult to track
+// the node or its contained nodes through all of the ir manipulation
+// involved in merging; it is pretty easy to uniquely identify the value
+// based on its uses. We can identify the value by its last use in the graph.
+// Values which do not have uses or which do not have a last use
+// outside of the subgraph to be merged into we do not need to track.
+struct ValueMapper {
+  ValueMapper(Node* n, AliasDb& db, size_t subgraph_num_outputs) {
+    last_uses_ = gatherLastUses(n->outputs());
+    subgraph_num_outputs_ = subgraph_num_outputs;
+    WithInsertPoint guard(n);
+    auto g = n->owningGraph();
+    // temporary node to put the aliasing properties of the node before its
+    // merged and destroyed
+    placeholder_node_ = g->insertNode(g->create(prim::Uninitialized, 0));
```

Comment:
nitpick: do we really need a node here? could we simply store all the values in a vector or will that confuse AA?

Reply:
Values are owned by the graph, if we store them in a vector but delete the node they will be dead references